### PR TITLE
Add option to not retry once daily limit has been reached

### PR DESCRIPTION
**Pull request by slukes**
This pull request addresses:
https://github.com/googlemaps/google-maps-services-java/issues/156

Which I realised there was already a pull request for after doing this work, however I note that the other pull request has not been merged - I hope this one will be.

- GeoApiContext now has an option to `failFastForDailyLimit` - i.e. to immediately throw an exception if a message containing `daily` comes with an error of  `OVER_QUERY_LIMIT`
- OverDailyLimitException has been changed to subclass OverQueryLimitException so that code, including my own which relies on the exception that is currently thrown can carry on working correctly.
- Unit test has been added

### CODE REVIEW

File: src/main/java/com/google/maps/internal/OkHttpPendingResult.java [hunk 1]
```diff
@@ -22,26 +22,12 @@
...
+import com.google.maps.errors.OverDailyLimitException;
 import com.google.maps.errors.OverQueryLimitException;
-import com.google.maps.model.AddressComponentType;
-import com.google.maps.model.AddressType;
-import com.google.maps.model.Distance;
-import com.google.maps.model.Duration;
-import com.google.maps.model.Fare;
-import com.google.maps.model.LatLng;
-import com.google.maps.model.LocationType;
+import com.google.maps.model.*;
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/internal/OkHttpPendingResult.java]: Please undo this change, Google style is to have all imports, not import *

File: src/main/java/com/google/maps/internal/OkHttpPendingResult.java [hunk 2]
```diff
@@ -22,26 +22,12 @@
...
-import com.google.maps.model.PriceLevel;
-import com.google.maps.model.TravelMode;
-
-import com.squareup.okhttp.Call;
-import com.squareup.okhttp.Callback;
-import com.squareup.okhttp.OkHttpClient;
-import com.squareup.okhttp.Request;
-import com.squareup.okhttp.Response;
-
+import com.squareup.okhttp.*;
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/internal/OkHttpPendingResult.java]: Ditto, undo

File: src/main/java/com/google/maps/internal/OkHttpPendingResult.java [hunk 3]
```diff
@@ -216,9 +205,9 @@ private T parseResponse(OkHttpPendingResult<T, R> request, Response response) th
 
     // Places Photo API special case
     if (contentType != null &&
-        contentType.startsWith("image") &&
-        responseClass == PhotoRequest.Response.class &&
-        response.code() == 200) {
+            contentType.startsWith("image") &&
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/internal/OkHttpPendingResult.java]: Please put back to 4 space indent: https://google.github.io/styleguide/javaguide.html#s4.2-block-indentation

File: src/main/java/com/google/maps/internal/OkHttpPendingResult.java [hunk 4]
```diff
@@ -227,31 +216,31 @@ private T parseResponse(OkHttpPendingResult<T, R> request, Response response) th
...
-        .registerTypeAdapter(TravelMode.class, new SafeEnumAdapter<TravelMode>(TravelMode.UNKNOWN))
-        .registerTypeAdapter(LocationType.class, new SafeEnumAdapter<LocationType>(LocationType.UNKNOWN))
-        .registerTypeAdapter(RatingType.class, new SafeEnumAdapter<RatingType>(RatingType.UNKNOWN))
-        .registerTypeAdapter(DayOfWeek.class, new DayOfWeekAdaptor())
-        .registerTypeAdapter(PriceLevel.class, new PriceLevelAdaptor())
-        .registerTypeAdapter(Instant.class, new InstantAdapter())
-        .registerTypeAdapter(LocalTime.class, new LocalTimeAdapter())
-        .setFieldNamingPolicy(fieldNamingPolicy)
-        .create();
+                    .registerTypeAdapter(DateTime.class, new DateTimeAdapter())
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/internal/OkHttpPendingResult.java]: Please put back to 4 space indent

File: src/main/java/com/google/maps/internal/OkHttpPendingResult.java [hunk 5]
```diff
@@ -227,31 +216,31 @@ private T parseResponse(OkHttpPendingResult<T, R> request, Response response) th
...
+                    .create();
 
     // Attempt to de-serialize before checking the HTTP status code, as there may be JSON in the
     // body that we can use to provide a more descriptive exception.
     try {
       resp = gson.fromJson(new String(bytes, "utf8"), responseClass);
     } catch (JsonSyntaxException e) {
       // Check HTTP status for a more suitable exception
-      if (!response.isSuccessful()) {
+      if (! response.isSuccessful()) {
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/internal/OkHttpPendingResult.java]: Undo change

File: src/main/java/com/google/maps/internal/OkHttpPendingResult.java [hunk 6]
```diff
@@ -278,12 +266,20 @@ private T parseResponse(OkHttpPendingResult<T, R> request, Response response) th
...
+               && cumulativeSleepTime < errorTimeOut;
+  }
+
   private byte[] getBytes(Response response) throws IOException {
     InputStream in = response.body().byteStream();
     ByteArrayOutputStream buffer = new ByteArrayOutputStream();
     int bytesRead;
     byte[] data = new byte[8192];
-    while ((bytesRead = in.read(data, 0, data.length)) != -1) {
+    while ((bytesRead = in.read(data, 0, data.length)) != - 1) {
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/internal/OkHttpPendingResult.java]: undo

File: src/test/java/com/google/maps/GeoApiContextTest.java [hunk 7]
```diff
@@ -36,6 +32,11 @@
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.*;
```
Comment (domesticmouse) [inline src/test/java/com/google/maps/GeoApiContextTest.java]: No star imports

**Comment by domesticmouse** [comment]
PTAL @markmcd

**Comment by slukes** [comment]
All code style issues mentioned above now fixed

File: src/main/java/com/google/maps/GeoApiContext.java [hunk 8]
```diff
@@ -167,7 +169,7 @@ public GeoApiContext(RequestHandler requestHandler) {
       hostName = baseUrlOverride;
     }
 
-    return requestHandler.handle(hostName, url.toString(), USER_AGENT, clazz, fieldNamingPolicy, errorTimeout);
+      return requestHandler.handle(hostName, url.toString(), USER_AGENT, clazz, fieldNamingPolicy, errorTimeout, failFastForDailyLimit);
```
Comment (domesticmouse) [inline src/main/java/com/google/maps/GeoApiContext.java]: What indenter are you using? This looks like it is no longer 2 space indent

**Comment by domesticmouse** [comment]
Hey @slukes, do you want to bring this PR up to date?

**Comment by domesticmouse** [comment]
Closing this out as @amyboyd's PR #189 has covered this. Thanks!
